### PR TITLE
[data] [3/3] [no_early_kickoff] Async iter batches e2e

### PR DESCRIPTION
**Pull request by amogkam**
Final PR for async iter_batches.

The new codepath is enabled for streaming execution. The old codepath is still accessible via a feature flag in `DatasetContext`.

Bulk execution still uses the old codepath by default.

This also deprecated `prefetch_batches` from `map_batches` since that is half baked and not entirely supported.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/stats.py [hunk 1]
```diff
@@ -639,6 +668,62 @@ class IterStatsSummary:
     iter_unknown_location: int
 
     def __str__(self) -> str:
+        if self.legacy_iter_batches:
+            return self.to_string_legacy()
+        else:
+            return self.to_string()
+
+    def to_string(self) -> str:
```
Comment (amogkam) [inline python/ray/data/_internal/stats.py:676]: new iterator stats output looks like this

```
Dataset iterator time breakdown:
* Total time user code is blocked: 24.0ms
* Total time in user code: 33.3ms
* Total time overall: 79.38s
* Num blocks local: 88
* Num blocks remote: 112
* Num blocks unknown location: 0
* Batch iteration time breakdown:
    * In ray.get(): 3.9ms min, 207.64ms max, 7.76ms avg, 1.55s total
    * In batch creation: 69.43us min, 36.3ms max, 4.31ms avg, 2.19s total
    * In batch formatting: 9.63ms min, 352.38ms max, 86.46ms avg, 43.92s total
    * In collate_fn: 41.41us min, 645.82us max, 102.03us avg, 51.83ms total
```

File: python/ray/data/_internal/block_batching/iter_batches.py [hunk 2]
```diff
@@ -1,47 +1,247 @@
...
-        - [Raylet] Batches 4 + 5 fetched to local object store memory
+    def _async_iter_batches(
+        block_refs: Iterator[Tuple[ObjectRef[Block], BlockMetadata]],
+    ) -> Iterator[DataBatch]:
 
-At any point in time there are prefetch_batches+1 batches in local heap memory.
-And the next set of prefetch_batches in local object store memory.
+        if prefetch_batches > 0:
+            # Step 1: Prefetch logical batches locally.
+            block_refs = prefetch_batches_locally(
```
Comment (ericl) [inline python/ray/data/_internal/block_batching/iter_batches.py:136]: Any way to always go through this code path (handle the zero case inside this function?) That would help avoid diverging code paths.

File: python/ray/data/_internal/block_batching/iter_batches.py [hunk 3]
```diff
@@ -1,47 +1,247 @@
...
+            except StopIteration:
+                break
+        with stats.iter_user_s.timer() if stats else nullcontext():
+            yield next_batch
+
+
+def _format_in_threadpool(
+    batch_iter: Iterator[Batch],
+    stats: DatasetStats,
+    batch_format: str = "default",
```
Comment (ericl) [inline python/ray/data/_internal/block_batching/iter_batches.py:196]:
```suggestion
    batch_format: Optional[str] = "default",
```

and elsewhere.

File: python/ray/data/_internal/block_batching/iter_batches.py [hunk 4]
```diff
@@ -1,47 +1,247 @@
...
+    collate_fn: Optional[Callable[[DataBatch], Any]] = None,
+    num_threadpool_workers: int = 0,
```
Comment (ericl) [inline python/ray/data/_internal/block_batching/iter_batches.py:198]: Shall we make all these args required?

File: python/ray/data/_internal/block_batching/iter_batches.py [hunk 5]
```diff
@@ -1,47 +1,247 @@
...
+        )
+
+        # Step 4b: Apply the collate function if applicable.
+        if collate_fn is not None:
+            formatted_batch_iter = collate(
+                formatted_batch_iter, collate_fn=collate_fn, stats=stats
+            )
+        yield from formatted_batch_iter
+
+    if num_threadpool_workers > 0:
```
Comment (ericl) [inline python/ray/data/_internal/block_batching/iter_batches.py:229]: Let's always use the thread pool?

File: python/ray/data/_internal/stats.py [hunk 6]
```diff
@@ -639,6 +668,62 @@ class IterStatsSummary:
...
+                fmt(self.block_time.get())
+            )
+        if self.user_time.get():
+            out += "* Total time in user code: {}\n".format(fmt(self.user_time.get()))
+        if self.total_time.get():
+            out += "* Total time overall: {}\n".format(fmt(self.total_time.get()))
+        out += "* Num blocks local: {}\n".format(self.iter_blocks_local)
+        out += "* Num blocks remote: {}\n".format(self.iter_blocks_remote)
+        out += "* Num blocks unknown location: {}\n".format(self.iter_unknown_location)
+        out += "* Batch iteration time breakdown:\n"
```
Comment (ericl) [inline python/ray/data/_internal/stats.py:690]:
```suggestion
        out += "* Batch iteration time breakdown (summed across prefetch threads):\n"
```

File: python/ray/data/_internal/block_batching/iter_batches.py (hunk 5 above)
Comment (amogkam) [reply python/ray/data/_internal/block_batching/iter_batches.py:229]: I don't this would be the correct behavior right? With a threadpool size of 1, we are prefetching 1 batch.

File: python/ray/data/_internal/block_batching/iter_batches.py (hunk 2 above)
Comment (amogkam) [reply python/ray/data/_internal/block_batching/iter_batches.py]: updated

File: python/ray/data/_internal/block_batching/iter_batches.py (hunk 3 above)
Comment (amogkam) [reply python/ray/data/_internal/block_batching/iter_batches.py]: ah sorry, updated

File: python/ray/data/_internal/block_batching/iter_batches.py (hunk 4 above)
Comment (amogkam) [reply python/ray/data/_internal/block_batching/iter_batches.py]: sure, updated

File: python/ray/data/_internal/stats.py (hunk 6 above)
Comment (amogkam) [reply python/ray/data/_internal/stats.py]: updated

File: python/ray/data/_internal/block_batching/iter_batches.py (hunk 5 above)
Comment (ericl) [reply python/ray/data/_internal/block_batching/iter_batches.py:223]: Got it. I'm a little bit wondering if we should make the prefetch default 1, but I guess that could double memory usage, is that right? On the other hand, bad perf by default is arguably worse.

tf.data seems to auto-tune their prefetch by default.

Comment (amogkam) [reply python/ray/data/_internal/block_batching/iter_batches.py:223]: Yeah we can make 1 the default, but you're right, this will double the memory usage.

Agree, auto-tuning is the ideal that we want.

File: python/ray/data/_internal/block_batching/iter_batches.py [hunk 7]
```diff
@@ -1,47 +1,241 @@
...
+    stats: Optional[DatasetStats] = None,
+    clear_block_after_read: bool = False,
+    batch_size: Optional[int] = None,
+    batch_format: Optional[str] = "default",
+    drop_last: bool = False,
+    collate_fn: Optional[Callable[[DataBatch], Any]] = None,
+    shuffle_buffer_min_size: Optional[int] = None,
+    shuffle_seed: Optional[int] = None,
+    ensure_copy: bool = False,
+    prefetch_batches: int = 0,
```
Comment (ericl) [inline python/ray/data/_internal/block_batching/iter_batches.py:48]:
```suggestion
    prefetch_batches: int = 1,
```

File: python/ray/data/_internal/stats.py [hunk 8]
```diff
@@ -34,21 +34,38 @@ class Timer:
...
         return self._value
 
+    def min(self) -> float:
+        return self._min
+
+    def max(self) -> float:
+        return self._max
+
+    def avg(self) -> float:
+        return self._value / self._total_count
```
Comment (ericl) [inline python/ray/data/_internal/stats.py:67]: Guard against NaN division?

File: python/ray/data/_internal/stats.py [hunk 9]
```diff
@@ -34,21 +34,38 @@ class Timer:
...
     def __init__(self):
         self._value: float = 0
+        self._min: float = float("inf")
+        self._max: float = 0
+        self._total_count: float = 0
 
     @contextmanager
     def timer(self) -> None:
-        time_start = time.perf_counter()
+        time_start = time.thread_time()
```
Comment (ericl) [inline python/ray/data/_internal/stats.py:43]: Hmm, why the change to thread (CPU) time? That can be unexpected when reported naively without the wall time context.

File: python/ray/data/dataset.py [hunk 10]
```diff
@@ -3406,8 +3430,12 @@ def to_tf(
             label_column: Columns that correspond to model targets. If this is a
                 string, the target data is a tensor. If this is a list, the target data
                 is a ``dict`` that maps column names to their tensor representation.
-            prefetch_blocks: The number of blocks to prefetch ahead of the
-                current block during the scan.
+            prefetch_batches: The number of batches to fetch ahead of the current batch
+                to fetch. If set to greater than 0, a separate threadpool will be used
+                to fetch the objects to the local node, format the batches, and apply
+                the collate_fn. Defaults to 1. This is still an alpha API. You can
```
Comment (ericl) [inline python/ray/data/dataset.py:3436]: Let's not document it as an alpha API (for those we'd call it experimental_prefetch_batches or something like that).

Comment (ericl) [reply python/ray/data/dataset.py:3436]: We can say this replaces the legacy prefetch_blocks option, which you can use by setting the use_legacy... etc..

File: python/ray/data/dataset_iterator.py [hunk 11]
```diff
@@ -97,7 +133,60 @@ def iter_batches(
...
+        if not context.use_streaming_executor:
+            # Always use legacy iter_batches for bulk executor.
+            use_legacy = True
+        else:
+            use_legacy = context.use_legacy_iter_batches
+
+        if prefetch_blocks > 0 and not use_legacy:
+            raise DeprecationWarning(
+                "`prefetch_blocks` arg is deprecated in Ray 2.4. Use "
+                "the`prefetch_batches` arg instead to specify the amount of "
```
Comment (ericl) [inline python/ray/data/dataset_iterator.py:147]:
```suggestion
                "the `prefetch_batches` arg instead to specify the amount of "
```

**Review by ericl — APPROVED** [review]

File: python/ray/data/_internal/block_batching/iter_batches.py (hunk 7 above)
Comment (amogkam) [reply python/ray/data/_internal/block_batching/iter_batches.py]: updated

File: python/ray/data/_internal/stats.py (hunk 8 above)
Comment (amogkam) [reply python/ray/data/_internal/stats.py]: updated

File: python/ray/data/_internal/stats.py (hunk 9 above)
Comment (amogkam) [reply python/ray/data/_internal/stats.py]: reverted back

File: python/ray/data/dataset.py (hunk 10 above)
Comment (amogkam) [reply python/ray/data/dataset.py]: updated

File: python/ray/data/dataset_iterator.py (hunk 11 above)
Comment (amogkam) [reply python/ray/data/dataset_iterator.py]: updated

**Comment by amogkam** [comment]
Failing tests are also failing on master, going to merge,
